### PR TITLE
Use mem, extract checks into smaller functions, and add unit tests

### LINKED ISSUE #11: Refactor
The code is quite messy right now. Would be nice to use [`mem`](https://github.com/sindresorhus/mem) on the exported function instead of assigning to the global `fullname` in every step. Also separate each check into a top level function, so we can easily mix and match checks for each platform, and maybe even do some checks in parallel using `Promise.race()`. I don't have time to do this right now, but could be an interesting exercise for anyone looking to contribute. PR welcome :)

### PR DESCRIPTION
**Pull request by canastro**
Fixes #11

### CODE REVIEW

**Comment by canastro** [comment]
Some remarks related with the unit tests:

- I had to include `mock-require` in order to be able to mock dependencies and reach most of the code in the unit tests. 
- I also had to clear require cache between tests in order for the new mocks to be applied
- The tests are running in serial because I am mocking environment and process variables

It was my first time mocking this kind of properties, If you know a more elegant way of doing this please share :)

File: index.js [hunk 1]
```diff
@@ -1,104 +1,110 @@
...
+			return user.fullname;
+		});
+}
 
-	if (envVar) {
-		fullname = envVar;
-		return Promise.resolve(fullname);
-	}
+function checkGit() {
+	return execa('git', ['config', '--global', 'user.name']).then(res => {
```
Comment (sindresorhus) [inline index.js]: You can use `execa.stdout()` here too.

File: index.js [hunk 2]
```diff
@@ -1,104 +1,110 @@
...
+}
 
+function checkOsaScript() {
+	return execa.stdout('osascript', ['-e', 'long user name of (system info)'])
+		.then(fullname => {
+			if (!fullname) {
+				throw new Error();
+			}
+
+			return fullname;
```
Comment (sindresorhus) [inline index.js]: Would be nicer to use a ternary here:

```js
.then(fullname => fullname || Promise.reject());
```

Same with other places.

File: index.js [hunk 3]
```diff
@@ -1,104 +1,110 @@
...
-		})
-		.catch(() => {
-			return execa.stdout('git', ['config', '--global', 'user.name'])
-				.then(stdout => {
-					fullname = stdout;
-					return fullname;
-				});
-		});
+function getFullName() {
+	return pAny([checkEnv(), checkAuthorName()])
```
Comment (sindresorhus) [inline index.js]: No point in using pAny here. `checkEnv` will always finish first. Just chain it.

File: test.js [hunk 4]
```diff
@@ -1,8 +1,376 @@
 import test from 'ava';
```
Comment (sindresorhus) [inline test.js]: Since all the tests are serial, you could use `import {serial as test} from 'ava';` and drop the serial property on the individual tests.

**Comment by canastro** [comment]
@sindresorhus updated PR with your comments :)

File: index.js [hunk 5]
```diff
@@ -1,104 +1,103 @@
...
+		resolve(fullname);
+	});
+}
 
-	const envVar = getEnvVar();
+function checkPasswd() {
+	return passwdUser()
+		.then(user => {
+			if (!user.fullname) {
+				throw new Error();
```
Comment (sindresorhus) [inline index.js]: You can use a ternary here too.

File: index.js [hunk 6]
```diff
@@ -1,104 +1,103 @@
...
+		.then(fullname => fullname || Promise.reject());
+}
+
+function checkWmic() {
+	return execa.stdout('wmic', ['useraccount', 'where', 'name="%username%"', 'get', 'fullname'])
+		.then(stdout => {
+			const fullname = stdout.replace('FullName', '');
+
+			if (!fullname) {
+				throw new Error();
```
Comment (sindresorhus) [inline index.js]: You can use a ternary here too.

File: index.js [hunk 7]
```diff
@@ -1,104 +1,103 @@
...
+			return fullname;
+		});
+}
 
+function checkGetEnt() {
+	return execa.shell('getent passwd $(whoami)').then(res => {
+		const fullname = (res.stdout.split(':')[4] || '').replace(/,.*/, '');
 		if (!fullname) {
-			return fallback().then(() => fullname);
+			throw new Error();
```
Comment (sindresorhus) [inline index.js]: You can use a ternary here too.

File: index.js [hunk 8]
```diff
@@ -1,104 +1,103 @@
...
 const execa = require('execa');
 const passwdUser = require('passwd-user');
-let fullname;
-let first = true;
+const pAny = require('p-any');
 
-function getEnvVar() {
-	const env = process.env;
+function checkEnv() {
+	return new Promise((resolve, reject) => {
```
Comment (sindresorhus) [inline index.js]: Can you use https://github.com/sindresorhus/p-try here instead?

File: index.js [hunk 9]
```diff
@@ -1,104 +1,103 @@
...
+		resolve(fullname);
+	});
 }
 
-module.exports = () => {
-	if (!first) {
-		return Promise.resolve(fullname);
-	}
+function checkAuthorName() {
+	return new Promise((resolve, reject) => {
```
Comment (sindresorhus) [inline index.js]: Can you use https://github.com/sindresorhus/p-try here instead?

File: test.js [hunk 10]
```diff
@@ -1,8 +1,359 @@
...
+	Object.defineProperty(process, 'platform', {
+		value: 'darwin'
+	});
+
+	mock('passwd-user', () => Promise.resolve({
+		fullname: 'TEST-PASSWD-FULL-NAME'
+	}));
+
+	// Required to delete require cache so the mock to passwd-usr is taken into consideration
+	delete require.cache[require.resolve('./')];
```
Comment (sindresorhus) [inline test.js]: Use https://github.com/sindresorhus/require-uncached

Same with the other places this is done.

**Comment by sindresorhus** [comment]
That was fast :)

Can you push new commits instead of rebasing? Makes it easier to review what's new.

**Comment by canastro** [comment]
Pull Request updated with a new commit 🦄

**Comment by sindresorhus** [comment]
Very good work on this PR @canastro 👍 ✨

**Pull request merged**